# NBA One Possession Fouling Analysis - Data Cleaning

## Imports

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
pd.options.display.max_columns = None

The data set was pulled from the link below. It includes play by play data for games between 2015 and mid 2021.

https://www.kaggle.com/datasets/schmadam97/nba-playbyplay-data-20182019/code

In [3]:
season2015 = pd.read_csv('C:/Users/kevin/Downloads/archive (1)/NBA_PBP_2015-16.csv')

In [4]:
season2015.head()

URL GameType                       Location  \
0  /boxscores/201510270ATL.html  regular  Philips Arena Atlanta Georgia   
1  /boxscores/201510270ATL.html  regular  Philips Arena Atlanta Georgia   
2  /boxscores/201510270ATL.html  regular  Philips Arena Atlanta Georgia   
3  /boxscores/201510270ATL.html  regular  Philips Arena Atlanta Georgia   
4  /boxscores/201510270ATL.html  regular  Philips Arena Atlanta Georgia   

              Date     Time WinningTeam  Quarter  SecLeft AwayTeam  \
0  October 27 2015  8:00 PM         DET        1      720      DET   
1  October 27 2015  8:00 PM         DET        1      701      DET   
2  October 27 2015  8:00 PM         DET        1      699      DET   
3  October 27 2015  8:00 PM         DET        1      697      DET   
4  October 27 2015  8:00 PM         DET        1      681      DET   

                                            AwayPlay  ...  FreeThrowNum  \
0  Jump ball: A. Drummond vs. A. Horford (E. lyas...  ...           NaN   
1  A. Drummond misses 2-pt layup from 1 ft (block...  ...           NaN   
2                                                NaN  ...           NaN   
3                                                NaN  ...           NaN   
4  M. Morris makes 2-pt jump shot from 13 ft (ass...  ...           NaN   

  EnterGame LeaveGame           TurnoverPlayer TurnoverType TurnoverCause  \
0       NaN       NaN                      NaN          NaN           NaN   
1       NaN       NaN                      NaN          NaN           NaN   
2       NaN       NaN                      NaN          NaN           NaN   
3       NaN       NaN  K. Bazemore - bazemke01     bad pass           NaN   
4       NaN       NaN                      NaN          NaN           NaN   

  TurnoverCauser       JumpballAwayPlayer      JumpballHomePlayer  \
0            NaN  A. Drummond - drumman01  A. Horford - horfoal01   
1            NaN                      NaN                     NaN   
2            NaN                      NaN                     NaN   
3            NaN                      NaN                     NaN   
4            NaN                      NaN                     NaN   

              JumpballPoss  
0  E. İlyasova - ilyaser01  
1                      NaN  
2                      NaN  
3                      NaN  
4                      NaN  

[5 rows x 40 columns]

In [5]:
season2015.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 601557 entries, 0 to 601556
Data columns (total 40 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   URL                 601557 non-null  object 
 1   GameType            601557 non-null  object 
 2   Location            601557 non-null  object 
 3   Date                601557 non-null  object 
 4   Time                601557 non-null  object 
 5   WinningTeam         601557 non-null  object 
 6   Quarter             601557 non-null  int64  
 7   SecLeft             601557 non-null  int64  
 8   AwayTeam            601557 non-null  object 
 9   AwayPlay            304900 non-null  object 
 10  AwayScore           601557 non-null  int64  
 11  HomeTeam            601557 non-null  object 
 12  HomePlay            296610 non-null  object 
 13  HomeScore           601557 non-null  int64  
 14  Shooter             222288 non-null  object 
 15  ShotType            222288 non-nul

In [6]:
season2015['Quarter'].unique()

array([1, 2, 3, 4, 5, 6, 7, 8], dtype=int64)

In [7]:
season2015[season2015['Quarter'] > 4].head()

URL GameType                      Location  \
10472  /boxscores/201510300ORL.html  regular  Amway Center Orlando Florida   
10473  /boxscores/201510300ORL.html  regular  Amway Center Orlando Florida   
10474  /boxscores/201510300ORL.html  regular  Amway Center Orlando Florida   
10475  /boxscores/201510300ORL.html  regular  Amway Center Orlando Florida   
10476  /boxscores/201510300ORL.html  regular  Amway Center Orlando Florida   

                  Date     Time WinningTeam  Quarter  SecLeft AwayTeam  \
10472  October 30 2015  7:00 PM         OKC        5      300      OKC   
10473  October 30 2015  7:00 PM         OKC        5      285      OKC   
10474  October 30 2015  7:00 PM         OKC        5      270      OKC   
10475  October 30 2015  7:00 PM         OKC        5      263      OKC   
10476  October 30 2015  7:00 PM         OKC        5      262      OKC   

                                                AwayPlay  ...  FreeThrowNum  \
10472  Jump ball: S. Adams vs. N. Vuevi (R. Westbrook...  ...           NaN   
10473       R. Westbrook makes 2-pt jump shot from 16 ft  ...           NaN   
10474                                                NaN  ...           NaN   
10475                                                NaN  ...           NaN   
10476                     Defensive rebound by K. Durant  ...           NaN   

      EnterGame LeaveGame  TurnoverPlayer TurnoverType TurnoverCause  \
10472       NaN       NaN             NaN          NaN           NaN   
10473       NaN       NaN             NaN          NaN           NaN   
10474       NaN       NaN             NaN          NaN           NaN   
10475       NaN       NaN             NaN          NaN           NaN   
10476       NaN       NaN             NaN          NaN           NaN   

      TurnoverCauser    JumpballAwayPlayer      JumpballHomePlayer  \
10472            NaN  S. Adams - adamsst01  N. Vučević - vucevni01   
10473            NaN                   NaN                     NaN   
10474            NaN                   NaN                     NaN   
10475            NaN                   NaN                     NaN   
10476            NaN                   NaN                     NaN   

                   JumpballPoss  
10472  R. Westbrook - westbru01  
10473                       NaN  
10474                       NaN  
10475                       NaN  
10476                       NaN  

[5 rows x 40 columns]

In [8]:
season2015.isna().sum()
#things that should be zero are zero

URL                        0
GameType                   0
Location                   0
Date                       0
Time                       0
WinningTeam                0
Quarter                    0
SecLeft                    0
AwayTeam                   0
AwayPlay              296657
AwayScore                  0
HomeTeam                   0
HomePlay              304947
HomeScore                  0
Shooter               379269
ShotType              379269
ShotOutcome           379269
ShotDist              379269
Assister              543345
Blocker               588526
FoulType              546577
Fouler                546577
Fouled                555585
Rebounder             464556
ReboundType           464556
ViolationPlayer       599235
ViolationType         599235
TimeoutTeam           583849
FreeThrowShooter      540037
FreeThrowOutcome      540037
FreeThrowNum          540037
EnterGame             542558
LeaveGame             542558
TurnoverPlayer        563897
TurnoverType  

In [9]:
season2016 = pd.read_csv('C:/Users/kevin/Downloads/archive (1)/NBA_PBP_2016-17.csv')

In [10]:
season2017 = pd.read_csv('C:/Users/kevin/Downloads/archive (1)/NBA_PBP_2017-18.csv')

In [11]:
season2018 = pd.read_csv('C:/Users/kevin/Downloads/archive (1)/NBA_PBP_2018-19.csv')

In [12]:
season2019 = pd.read_csv('C:/Users/kevin/Downloads/archive (1)/NBA_PBP_2019-20.csv')

In [13]:
season2020 = pd.read_csv('C:/Users/kevin/Downloads/archive (1)/NBA_PBP_2020-21.csv')

Each of the individual seasons are its own data set so I concatenated them to work with one massive dataframe with all plays from all games

In [64]:
seasons = pd.concat([season2015, season2016, season2017, season2018, season2019, season2020], axis=0, ignore_index = True)

In [39]:
seasons.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3040524 entries, 0 to 3040523
Data columns (total 41 columns):
 #   Column              Dtype  
---  ------              -----  
 0   URL                 object 
 1   GameType            object 
 2   Location            object 
 3   Date                object 
 4   Time                object 
 5   WinningTeam         object 
 6   Quarter             int64  
 7   SecLeft             int64  
 8   AwayTeam            object 
 9   AwayPlay            object 
 10  AwayScore           int64  
 11  HomeTeam            object 
 12  HomePlay            object 
 13  HomeScore           int64  
 14  Shooter             object 
 15  ShotType            object 
 16  ShotOutcome         object 
 17  ShotDist            float64
 18  Assister            object 
 19  Blocker             object 
 20  FoulType            object 
 21  Fouler              object 
 22  Fouled              object 
 23  Rebounder           object 
 24  ReboundType         obje

In [40]:
seasons.columns

Index(['URL', 'GameType', 'Location', 'Date', 'Time', 'WinningTeam', 'Quarter',
       'SecLeft', 'AwayTeam', 'AwayPlay', 'AwayScore', 'HomeTeam', 'HomePlay',
       'HomeScore', 'Shooter', 'ShotType', 'ShotOutcome', 'ShotDist',
       'Assister', 'Blocker', 'FoulType', 'Fouler', 'Fouled', 'Rebounder',
       'ReboundType', 'ViolationPlayer', 'ViolationType', 'TimeoutTeam',
       'FreeThrowShooter', 'FreeThrowOutcome', 'FreeThrowNum', 'EnterGame',
       'LeaveGame', 'TurnoverPlayer', 'TurnoverType', 'TurnoverCause',
       'TurnoverCauser', 'JumpballAwayPlayer', 'JumpballHomePlayer',
       'JumpballPoss', 'Unnamed: 40'],
      dtype='object')

In [41]:
season2015.columns

Index(['URL', 'GameType', 'Location', 'Date', 'Time', 'WinningTeam', 'Quarter',
       'SecLeft', 'AwayTeam', 'AwayPlay', 'AwayScore', 'HomeTeam', 'HomePlay',
       'HomeScore', 'Shooter', 'ShotType', 'ShotOutcome', 'ShotDist',
       'Assister', 'Blocker', 'FoulType', 'Fouler', 'Fouled', 'Rebounder',
       'ReboundType', 'ViolationPlayer', 'ViolationType', 'TimeoutTeam',
       'FreeThrowShooter', 'FreeThrowOutcome', 'FreeThrowNum', 'EnterGame',
       'LeaveGame', 'TurnoverPlayer', 'TurnoverType', 'TurnoverCause',
       'TurnoverCauser', 'JumpballAwayPlayer', 'JumpballHomePlayer',
       'JumpballPoss'],
      dtype='object')

In [42]:
seasons['Unnamed: 40'].unique()

array([nan])

In [65]:
seasons.drop('Unnamed: 40', axis=1, inplace=True)

## Data Cleaning

Wanted to first shrink the dataset by cutting out the first three quarters. Our analysis is contained to end of game situations anyways.

In [66]:
seasons = seasons[seasons['Quarter'] > 3]

In [45]:
seasons['Quarter'].unique()

array([4, 5, 6, 7, 8], dtype=int64)

In [46]:
seasons['SecLeft'].dtype

dtype('int64')

The next thing I did to filter down the data was remove all rows in which there was more than 34 seconds left. At over 34 seconds left, I can safely assume that no one is committing take fouls down only one possession.

In [67]:
seasons = seasons[seasons['SecLeft'] <= 34]

In [49]:
seasons.tail(10)

URL GameType  \
3039615  /boxscores/202101110ATL.html  regular   
3040047  /boxscores/202101100BRK.html  regular   
3040048  /boxscores/202101100BRK.html  regular   
3040049  /boxscores/202101100BRK.html  regular   
3040518  /boxscores/202101150UTA.html  regular   
3040519  /boxscores/202101150UTA.html  regular   
3040520  /boxscores/202101150UTA.html  regular   
3040521  /boxscores/202101150UTA.html  regular   
3040522  /boxscores/202101150UTA.html  regular   
3040523  /boxscores/202101150UTA.html  regular   

                                            Location             Date  \
3039615             State Farm Arena Atlanta Georgia  January 11 2021   
3040047            Barclays Center Brooklyn New York  January 10 2021   
3040048            Barclays Center Brooklyn New York  January 10 2021   
3040049            Barclays Center Brooklyn New York  January 10 2021   
3040518  Vivint Smart Home Arena Salt Lake City Utah  January 15 2021   
3040519  Vivint Smart Home Arena Salt Lake City Utah  January 15 2021   
3040520  Vivint Smart Home Arena Salt Lake City Utah  January 15 2021   
3040521  Vivint Smart Home Arena Salt Lake City Utah  January 15 2021   
3040522  Vivint Smart Home Arena Salt Lake City Utah  January 15 2021   
3040523  Vivint Smart Home Arena Salt Lake City Utah  January 15 2021   

            Time WinningTeam  Quarter  SecLeft AwayTeam  \
3039615  7:30 PM         ATL        4        0      PHI   
3040047  6:00 PM         OKC        4       11      OKC   
3040048  6:00 PM         OKC        4        0      OKC   
3040049  6:00 PM         OKC        4        0      OKC   
3040518  9:00 PM         UTA        4       34      ATL   
3040519  9:00 PM         UTA        4       15      ATL   
3040520  9:00 PM         UTA        4       13      ATL   
3040521  9:00 PM         UTA        4       10      ATL   
3040522  9:00 PM         UTA        4        0      ATL   
3040523  9:00 PM         UTA        4        0      ATL   

                                                  AwayPlay  AwayScore  \
3039615                                        End of Game         94   
3040047                      Turnover by Team (shot clock)        129   
3040048                                 End of 4th quarter        129   
3040049                                        End of Game        129   
3040518  T. Snell makes 3-pt jump shot from 27 ft (assi...         92   
3040519                                                NaN         92   
3040520                                                NaN         92   
3040521                                                NaN         92   
3040522                                 End of 4th quarter         92   
3040523                                        End of Game         92   

        HomeTeam                                           HomePlay  \
3039615      ATL                                                NaN   
3040047      BRK                                                NaN   
3040048      BRK                                                NaN   
3040049      BRK                                                NaN   
3040518      UTA                                                NaN   
3040519      UTA            S. Harrison misses 2-pt layup from 3 ft   
3040520      UTA                   Offensive rebound by U. Azubuike   
3040521      UTA  Turnover by U. Azubuike (bad pass; steal by O....   
3040522      UTA                                                NaN   
3040523      UTA                                                NaN   

         HomeScore                  Shooter        ShotType ShotOutcome  \
3039615        112                      NaN             NaN         NaN   
3040047        116                      NaN             NaN         NaN   
3040048        116                      NaN             NaN         NaN   
3040049        116                      NaN             NaN         NaN   
3040518        116     T. Snell - snellto01  3-pt jump shot        m

The following code is a batch I ended up using much more so I realized this later and made a function for it. It essentially checks the amount of unique games in the dataset to show if our steps are doing any filtering or not.

In [68]:
# group the DataFrame by 'Date' and 'Location'
grouped = seasons.groupby(['URL'])

# get the number of groups by taking the length of the grouped DataFrame
num_groups = len(grouped)

# print the number of groups
print("Number of groups:", num_groups)

Number of groups: 6600


The below code groups the dataframe by URL, which is each unique game. Then it filters out until there are only games remaining in which the first row depicts a play by play situation that represents a one score game. If you remember, the first row will be the first play below 34 seconds left in the game.

In [ ]:
#Recopied for GitHub
# create a new column 'score_diff' that contains the difference between the two score columns
seasons['score_diff'] = seasons.apply(lambda x: abs(x['AwayScore'] - x['HomeScore']), axis=1)

# group the DataFrame by 'URL'
grouped = seasons.groupby('URL')
qualified_groups = 0
counter = 0

# create a new empty DataFrame called close_games
close_games = pd.DataFrame(columns=seasons.columns)

# iterate over the grouped DataFrame
for name, group in grouped:
    # get the first row of the group
    first_row = group.iloc[0,:]
    counter += 1
    
    
    # check if the score_diff of the first row is either 3, 2, or 1
    if first_row['score_diff'] in [3, 2, 1]:
        # if it is, add the entire group to close_games
        close_games = pd.concat([close_games, group], ignore_index=True)
        qualified_groups += 1
        
print(qualified_groups) 

In [59]:
close_games

URL GameType  \
0      /boxscores/201510270CHI.html  regular   
1      /boxscores/201510270CHI.html  regular   
2      /boxscores/201510270CHI.html  regular   
3      /boxscores/201510270CHI.html  regular   
4      /boxscores/201510270CHI.html  regular   
...                             ...      ...   
27685  /boxscores/202101200MIN.html  regular   
27686  /boxscores/202101200MIN.html  regular   
27687  /boxscores/202101200MIN.html  regular   
27688  /boxscores/202101200MIN.html  regular   
27689  /boxscores/202101200MIN.html  regular   

                                  Location             Date     Time  \
0           United Center Chicago Illinois  October 27 2015  8:00 PM   
1           United Center Chicago Illinois  October 27 2015  8:00 PM   
2           United Center Chicago Illinois  October 27 2015  8:00 PM   
3           United Center Chicago Illinois  October 27 2015  8:00 PM   
4           United Center Chicago Illinois  October 27 2015  8:00 PM   
...                                    ...              ...      ...   
27685  Target Center Minneapolis Minnesota  January 20 2021  8:00 PM   
27686  Target Center Minneapolis Minnesota  January 20 2021  8:00 PM   
27687  Target Center Minneapolis Minnesota  January 20 2021  8:00 PM   
27688  Target Center Minneapolis Minnesota  January 20 2021  8:00 PM   
27689  Target Center Minneapolis Minnesota  January 20 2021  8:00 PM   

      WinningTeam Quarter SecLeft AwayTeam  \
0             CHI       4      33      CLE   
1             CHI       4      10      CLE   
2             CHI       4      10      CLE   
3             CHI       4      10      CLE   
4             CHI       4       4      CLE   
...           ...     ...     ...      ...   
27685         ORL       4       1      ORL   
27686         ORL       4       0      ORL   
27687         ORL       4       0      ORL   
27688         ORL       4       0      ORL   
27689         ORL       4       0      ORL   

                                                AwayPlay AwayScore HomeTeam  \
0      K. Love makes 3-pt jump shot from 25 ft (assis...        95      CHI   
1                                                    NaN        95      CHI   
2                       Defensive rebound by T. Thompson        95      CHI   
3                                 Cleveland full timeout        95      CHI   
4      L. James misses 2-pt layup from 2 ft (block by...        95      CHI   
...                                                  ...       ...      ...   
27685                    Defensive rebound by C. Anthony        94      MIN   
27686         C. Anthony makes 3-pt jump shot from 25 ft        97      MIN   
27687            Instant Replay (Request: Ruling Stands)        97      MIN   
27688                                 End of 4th quarter        97      MIN   
27689                                        End of Game        97      MIN   

                                       HomePlay HomeScore  \
0                                           NaN        97   
1      D. Rose misses 2-pt jump shot from 20 ft        97   
2                                           NaN        97   
3                                           NaN        97   
4                                           NaN        97   
...                                         ...       ...   
27685                                       NaN        96   
27686                                       NaN        96   
27687                                       NaN        96   
27688                                       NaN        96   
27689                                       NaN        96   

                      Shooter        ShotType ShotOutcome  ShotDist  \
0          K. Love - loveke01  3-pt jump shot        make      25.0   
1          D. Rose - rosede01  2-pt jump shot        miss      20.0   
2                         NaN             NaN         NaN       NaN   
3                         NaN             NaN         NaN       NaN   


The below code removes any code that goes to overtime. This makes my analysis less comprehensive but makes the code less complex. If I kept overtime games, using the winning team column would no longer be a good representation of whether the take foul worked. This is because the strategy could work sending the game to overtime, but then they could lose in overtime, which would still show as a loss in a take foul game unless I personalized the code further.

In [70]:
# group the DataFrame by 'URL'
grouped = close_games.groupby('URL')

# initialize a variable to keep track of the number of groups that meet the criteria
qualified_groups = 0

# create a new DataFrame to hold the qualified groups
close_games_filtered = pd.DataFrame(columns=close_games.columns)

# iterate over the grouped DataFrame
for name, group in grouped:
    # check if all the rows in the group have 'Quarter' equal to 4
    if (group['Quarter'] == 4).all():
        # if they do, keep the entire group
        close_games_filtered = pd.concat([close_games_filtered, group], ignore_index=True)
        qualified_groups += 1

print(qualified_groups)
# overwrite close_games with close_games_filtered


985


In [71]:
close_games_filtered[close_games_filtered['Quarter'] > 4]

Empty DataFrame
Columns: [URL, GameType, Location, Date, Time, WinningTeam, Quarter, SecLeft, AwayTeam, AwayPlay, AwayScore, HomeTeam, HomePlay, HomeScore, Shooter, ShotType, ShotOutcome, ShotDist, Assister, Blocker, FoulType, Fouler, Fouled, Rebounder, ReboundType, ViolationPlayer, ViolationType, TimeoutTeam, FreeThrowShooter, FreeThrowOutcome, FreeThrowNum, EnterGame, LeaveGame, TurnoverPlayer, TurnoverType, TurnoverCause, TurnoverCauser, JumpballAwayPlayer, JumpballHomePlayer, JumpballPoss, score_diff]
Index: []

In [74]:
close_games_filtered.sample()

URL GameType                       Location  \
14946  /boxscores/201903280SAS.html  regular  AT&T Center San Antonio Texas   

                Date     Time WinningTeam Quarter SecLeft AwayTeam  \
14946  March 28 2019  8:30 PM         SAS       4       5      CLE   

                                           AwayPlay AwayScore HomeTeam  \
14946  J. Clarkson misses 3-pt jump shot from 24 ft       110      SAS   

      HomePlay HomeScore                  Shooter        ShotType ShotOutcome  \
14946      NaN       114  J. Clarkson - clarkjo01  3-pt jump shot        miss   

       ShotDist Assister Blocker FoulType Fouler Fouled Rebounder ReboundType  \
14946      24.0      NaN     NaN      NaN    NaN    NaN       NaN         NaN   

      ViolationPlayer ViolationType TimeoutTeam FreeThrowShooter  \
14946             NaN           NaN         NaN              NaN   

      FreeThrowOutcome FreeThrowNum EnterGame LeaveGame TurnoverPlayer  \
14946              NaN          NaN       NaN       NaN            NaN   

      TurnoverType TurnoverCause TurnoverCauser JumpballAwayPlayer  \
14946          NaN           NaN            NaN                NaN   

      JumpballHomePlayer JumpballPoss score_diff  
14946                NaN          NaN          4

In [87]:
close_games_filtered

URL GameType  \
0      /boxscores/201510270CHI.html  regular   
1      /boxscores/201510270CHI.html  regular   
2      /boxscores/201510270CHI.html  regular   
3      /boxscores/201510270CHI.html  regular   
4      /boxscores/201510270CHI.html  regular   
...                             ...      ...   
19068  /boxscores/202101200MIN.html  regular   
19069  /boxscores/202101200MIN.html  regular   
19070  /boxscores/202101200MIN.html  regular   
19071  /boxscores/202101200MIN.html  regular   
19072  /boxscores/202101200MIN.html  regular   

                                  Location             Date     Time  \
0           United Center Chicago Illinois  October 27 2015  8:00 PM   
1           United Center Chicago Illinois  October 27 2015  8:00 PM   
2           United Center Chicago Illinois  October 27 2015  8:00 PM   
3           United Center Chicago Illinois  October 27 2015  8:00 PM   
4           United Center Chicago Illinois  October 27 2015  8:00 PM   
...                                    ...              ...      ...   
19068  Target Center Minneapolis Minnesota  January 20 2021  8:00 PM   
19069  Target Center Minneapolis Minnesota  January 20 2021  8:00 PM   
19070  Target Center Minneapolis Minnesota  January 20 2021  8:00 PM   
19071  Target Center Minneapolis Minnesota  January 20 2021  8:00 PM   
19072  Target Center Minneapolis Minnesota  January 20 2021  8:00 PM   

      WinningTeam Quarter SecLeft AwayTeam  \
0             CHI       4      33      CLE   
1             CHI       4      10      CLE   
2             CHI       4      10      CLE   
3             CHI       4      10      CLE   
4             CHI       4       4      CLE   
...           ...     ...     ...      ...   
19068         ORL       4       1      ORL   
19069         ORL       4       0      ORL   
19070         ORL       4       0      ORL   
19071         ORL       4       0      ORL   
19072         ORL       4       0      ORL   

                                                AwayPlay AwayScore HomeTeam  \
0      K. Love makes 3-pt jump shot from 25 ft (assis...        95      CHI   
1                                                    NaN        95      CHI   
2                       Defensive rebound by T. Thompson        95      CHI   
3                                 Cleveland full timeout        95      CHI   
4      L. James misses 2-pt layup from 2 ft (block by...        95      CHI   
...                                                  ...       ...      ...   
19068                    Defensive rebound by C. Anthony        94      MIN   
19069         C. Anthony makes 3-pt jump shot from 25 ft        97      MIN   
19070            Instant Replay (Request: Ruling Stands)        97      MIN   
19071                                 End of 4th quarter        97      MIN   
19072                                        End of Game        97      MIN   

                                       HomePlay HomeScore  \
0                                           NaN        97   
1      D. Rose misses 2-pt jump shot from 20 ft        97   
2                                           NaN        97   
3                                           NaN        97   
4                                           NaN        97   
...                                         ...       ...   
19068                                       NaN        96   
19069                                       NaN        96   
19070                                       NaN        96   
19071                                       NaN        96   
19072                                       NaN        96   

                      Shooter        ShotType ShotOutcome  ShotDist  \
0          K. Love - loveke01  3-pt jump shot        make      25.0   
1          D. Rose - rosede01  2-pt jump shot        miss      20.0   
2                         NaN             NaN         NaN       NaN   
3                         NaN             NaN         NaN       NaN   


In [92]:
# group the DataFrame by 'Date' and 'Location'
grouped = close_games_filtered.groupby(['URL'])

# get the number of groups by taking the length of the grouped DataFrame
num_groups = len(grouped)

# print the number of groups
print("Number of groups:", num_groups)

Number of groups: 985


The next piece of filtering I did was to remove games in which the first row remaining (under 34 seconds) was also under 24 seconds. This is because if there is less than 24 seconds left in the game, the losing team has no choice but to foul.

In [89]:
# group the DataFrame by 'URL'
grouped = close_games_filtered.groupby('URL')
qualified_groups = 0

# create a new empty DataFrame called close_games
secs24 = pd.DataFrame(columns=close_games_filtered.columns)

# iterate over the grouped DataFrame
for name, group in grouped:
    # get the first row of the group
    first_row = group.iloc[0,:]
    
    # check if the first row of the group has a 'SecLeft' value of 24 or less
    if first_row['SecLeft'] > 24:
        # if it is, add the entire group to close_games
        secs24 = pd.concat([secs24, group], ignore_index=True)
        qualified_groups += 1
        
print(qualified_groups)

728


In [91]:
# group the DataFrame by 'Date' and 'Location'
grouped = secs24.groupby(['URL'])

# get the number of groups by taking the length of the grouped DataFrame
num_groups = len(grouped)

# print the number of groups
print("Number of groups:", num_groups)

Number of groups: 728


Now comes the more complex code in which I try to whittle down the games exactly to the possibility of a take foul down one possession. This means my goal is to get games in which the first row is a NaN for the winning team play (meaning the losing just scored or turned the ball over and the winning team has the ball again) or the winning team play is a rebound (again meaning they have the ball again while up).

I begin by trying to remove the edge cases in this logic, though. For this first edge case: If the losing team gets an offensive rebound, the winning team's play will be NaN but the losing team still has the ball. As you can see, five instances of this are removed (728 groups - 723 groups).

In [94]:
# group the DataFrame by 'Date' and 'Location'
grouped = secs24.groupby('URL')

# create an empty list to store the qualified groups that will be removed from 'seasons'
disqualified_groups = []

# iterate over each group
for name, group in grouped:
    first_row = group.iloc[0,:]
    if (not pd.isnull(first_row['AwayPlay']) and 'Offensive rebound' in first_row['AwayPlay'] and first_row['AwayScore'] < first_row['HomeScore']) or (not pd.isnull(first_row['HomePlay']) and 'Offensive rebound' in first_row['HomePlay'] and first_row['HomeScore'] < first_row['AwayScore']):
        disqualified_groups.append(group)

# concatenate the groups and drop the rows from 'seasons'
noOffReb = secs24.drop(pd.concat(disqualified_groups).index)

In [95]:
# group the DataFrame by 'Date' and 'Location'
grouped = noOffReb.groupby(['URL'])

# get the number of groups by taking the length of the grouped DataFrame
num_groups = len(grouped)

# print the number of groups
print("Number of groups:", num_groups)

Number of groups: 723


The next edge case to be removed is if the winning team is fouled. I didn't want them to get free throws and then no longer be in the one possession stipulation. I made sure to write an exception that still keeps the rows if the foul is a take foul.

In [98]:
grouped = noOffReb.groupby('URL')
disqualified_groups = []

for name, group in grouped:
    first_row = group.iloc[0,:]
    if ((not pd.isnull(first_row['AwayPlay']) and 'foul' in str(first_row['AwayPlay']) and first_row['AwayScore'] > first_row['HomeScore']) or (not pd.isnull(first_row['HomePlay']) and 'foul' in str(first_row['HomePlay']) and first_row['HomeScore'] > first_row['AwayScore'])) and ('take foul' not in str(first_row['AwayPlay']) and 'take foul' not in str(first_row['HomePlay'])):
        disqualified_groups.append(group)
        
noPF = noOffReb.drop(pd.concat(disqualified_groups).index)

In [99]:
# group the DataFrame by 'Date' and 'Location'
grouped = noPF.groupby(['URL'])

# get the number of groups by taking the length of the grouped DataFrame
num_groups = len(grouped)

# print the number of groups
print("Number of groups:", num_groups)

Number of groups: 688


As you can see, I've used the code to check the number of groups a few times, so I just made a function for it.

In [3]:
def check_groups(df):
    # group the DataFrame by 'Date' and 'Location'
    grouped = df.groupby(['URL'])

    # get the number of groups by taking the length of the grouped DataFrame
    num_groups = len(grouped)

    # print the number of groups
    print("Number of groups:", num_groups)

I then wrote code that removes personal foul rows in which the winning team commits a foul. This is necessary because the first row play for the winning team would be NaN as the foul shows up in the column of the team drawing the foul, but the winning team would still have the ball. This means the losing team is not in a fouling situation.

In [6]:
grouped = noPF.groupby('URL')
disqualified_groups = []

for name, group in grouped:
    first_row = group.iloc[0,:]
    if ((not pd.isnull(first_row['AwayPlay']) and 'foul' in str(first_row['AwayPlay']) and first_row['HomeScore'] > first_row['AwayScore'])):
        disqualified_groups.append(group)
    elif ((not pd.isnull(first_row['HomePlay']) and 'foul' in str(first_row['HomePlay']) and first_row['AwayScore'] > first_row['HomeScore'])):
        disqualified_groups.append(group)
        
noWinFouls = noPF.drop(pd.concat(disqualified_groups).index)

In [7]:
check_groups(noWinFouls)

Number of groups: 648


The next edge case is removing games in which the first row is a timeout from the losing team. Similarly, the winning team's play column for the first row would show NaN, but the losing team still has the ball, meaning they will not commit a take foul.

In [8]:
grouped = noWinFouls.groupby('URL')
disqualified_groups = []

for name, group in grouped:
    first_row = group.iloc[0,:]
    if ((not pd.isnull(first_row['AwayPlay']) and 'timeout' in str(first_row['AwayPlay']) and first_row['HomeScore'] > first_row['AwayScore'])):
        disqualified_groups.append(group)
    elif ((not pd.isnull(first_row['HomePlay']) and 'timeout' in str(first_row['HomePlay']) and first_row['AwayScore'] > first_row['HomeScore'])):
        disqualified_groups.append(group)
        
noTimeout = noWinFouls.drop(pd.concat(disqualified_groups).index)

In [9]:
check_groups(noTimeout)

Number of groups: 633


I also removed games in which the first play remaining was a replay. This is because it would be ambiguous who has the ball without reading the play for the specific game.

In [15]:
grouped = noTimeout.groupby('URL')
disqualified_groups = []

for name, group in grouped:
    first_row = group.iloc[0,:]
    if (not pd.isnull(first_row['AwayPlay']) and 'Replay' in str(first_row['AwayPlay'])):
        disqualified_groups.append(group)
    elif (not pd.isnull(first_row['HomePlay']) and 'Replay' in str(first_row['HomePlay'])):
        disqualified_groups.append(group)


noReplays = noTimeout.drop(pd.concat(disqualified_groups).index)

In [16]:
check_groups(noReplays)

Number of groups: 632


I then removed games in which the first play is a substitution play. This is for a similar reason. I would not know who had the ball. Luckily there are only two instances.

In [17]:
grouped = noReplays.groupby('URL')
disqualified_groups = []

for name, group in grouped:
    first_row = group.iloc[0,:]
    if (not pd.isnull(first_row['AwayPlay']) and 'enters the game' in str(first_row['AwayPlay'])):
        disqualified_groups.append(group)
    elif (not pd.isnull(first_row['HomePlay']) and 'enters the game' in str(first_row['HomePlay'])):
        disqualified_groups.append(group)

noSubs = noReplays.drop(pd.concat(disqualified_groups).index)

In [18]:
check_groups(noSubs)

Number of groups: 630


Finally, all edge cases are removed. The below code creates the final list of games in which the first play row for the winning team is either a NaN or a rebound, meaning they have the ball with 24-34 seconds remaining, and the losing team is down by one possession with the decision of whether or not to foul to extend the game.

In [19]:
grouped = noSubs.groupby('URL')
qualified_groups = []

for name, group in grouped:
    first_row = group.iloc[0,:]
    if ((pd.isnull(first_row['AwayPlay']) or ('rebound' in str(first_row['AwayPlay']))) and first_row['AwayScore'] > first_row['HomeScore']):
        qualified_groups.append(group)
    elif ((pd.isnull(first_row['HomePlay']) or ('rebound' in str(first_row['HomePlay']))) and first_row['HomeScore'] > first_row['AwayScore']):
        qualified_groups.append(group)

final_games = pd.concat(qualified_groups)

In [20]:
check_groups(final_games)

Number of groups: 332


In [27]:
final_games

URL GameType  \
0      /boxscores/201510270CHI.html  regular   
1      /boxscores/201510270CHI.html  regular   
2      /boxscores/201510270CHI.html  regular   
3      /boxscores/201510270CHI.html  regular   
4      /boxscores/201510270CHI.html  regular   
...                             ...      ...   
13627  /boxscores/202101200MIN.html  regular   
13628  /boxscores/202101200MIN.html  regular   
13629  /boxscores/202101200MIN.html  regular   
13630  /boxscores/202101200MIN.html  regular   
13631  /boxscores/202101200MIN.html  regular   

                                  Location             Date     Time  \
0           United Center Chicago Illinois  October 27 2015  8:00 PM   
1           United Center Chicago Illinois  October 27 2015  8:00 PM   
2           United Center Chicago Illinois  October 27 2015  8:00 PM   
3           United Center Chicago Illinois  October 27 2015  8:00 PM   
4           United Center Chicago Illinois  October 27 2015  8:00 PM   
...                                    ...              ...      ...   
13627  Target Center Minneapolis Minnesota  January 20 2021  8:00 PM   
13628  Target Center Minneapolis Minnesota  January 20 2021  8:00 PM   
13629  Target Center Minneapolis Minnesota  January 20 2021  8:00 PM   
13630  Target Center Minneapolis Minnesota  January 20 2021  8:00 PM   
13631  Target Center Minneapolis Minnesota  January 20 2021  8:00 PM   

      WinningTeam  Quarter  SecLeft AwayTeam  \
0             CHI        4       33      CLE   
1             CHI        4       10      CLE   
2             CHI        4       10      CLE   
3             CHI        4       10      CLE   
4             CHI        4        4      CLE   
...           ...      ...      ...      ...   
13627         ORL        4        1      ORL   
13628         ORL        4        0      ORL   
13629         ORL        4        0      ORL   
13630         ORL        4        0      ORL   
13631         ORL        4        0      ORL   

                                                AwayPlay  AwayScore HomeTeam  \
0      K. Love makes 3-pt jump shot from 25 ft (assis...         95      CHI   
1                                                    NaN         95      CHI   
2                       Defensive rebound by T. Thompson         95      CHI   
3                                 Cleveland full timeout         95      CHI   
4      L. James misses 2-pt layup from 2 ft (block by...         95      CHI   
...                                                  ...        ...      ...   
13627                    Defensive rebound by C. Anthony         94      MIN   
13628         C. Anthony makes 3-pt jump shot from 25 ft         97      MIN   
13629            Instant Replay (Request: Ruling Stands)         97      MIN   
13630                                 End of 4th quarter         97      MIN   
13631                                        End of Game         97      MIN   

                                       HomePlay  HomeScore  \
0                                           NaN         97   
1      D. Rose misses 2-pt jump shot from 20 ft         97   
2                                           NaN         97   
3                                           NaN         97   
4                                           NaN         97   
...                                         ...        ...   
13627                                       NaN         96   
13628                                       NaN         96   
13629                                       NaN         96   
13630                                       NaN         96   
13631                                       NaN         96   

                      Shooter        ShotType ShotOutcome  ShotDist  \
0          K. Love - loveke01  3-pt jump shot        make      25.0   
1          D. Rose - rosede01  2-pt jump shot        miss      20.0   
2                         NaN             NaN         NaN       NaN   
3                      

In [60]:
final_games.to_csv('C:/Users/kevin/Downloads/final games.csv')

The below code creates a new column called 'take_foul'. It goes through the rest of the plays for each game, and if a take foul is committed by the losing team with more than 24 seconds remaining, the game would be 'True'.

In [41]:
grouped = final_games.groupby('URL')

for name, group in grouped:
    if (((group['HomeScore'] > group['AwayScore']).any() & (group['HomePlay'].str.contains('take foul') & (group['SecLeft'] > 24)).any()) |
        ((group['HomeScore'] < group['AwayScore']).any() & (group['AwayPlay'].str.contains('take foul') & (group['SecLeft'] > 24)).any())):
        final_games.loc[group.index, 'take_foul'] = True

All the games in which the situation is True is put in the dataframe take_foul_games while all games in which the situation is False is put in the dataframe no_take_foul

In [44]:
take_foul_games = final_games[final_games['take_foul'] == True]
no_take_foul = final_games[final_games['take_foul'] == False]

In [50]:
check_groups(take_foul_games)

Number of groups: 12


In [51]:
check_groups(no_take_foul)

Number of groups: 320


I then decided I wanted to repurpose this dataset to also take a quick look at teams committing take fouls when winning. This could either be because they are up 3 and want to prevent a game tying attempt, or they have a foul to give and want to make the losing team reset their action.

The code below does a similar thing to the take_foul column code below. This code assigns a game True if the winning team commits a take foul at any point (no seconds left restriction).

In [52]:
final_games['winning_take_foul'] = False

grouped = final_games.groupby('URL')

for name, group in grouped:
    if (((group['HomeScore'] > group['AwayScore']).any() & (group['AwayPlay'].str.contains('take foul')).any()) |
        ((group['HomeScore'] < group['AwayScore']).any() & (group['HomePlay'].str.contains('take foul')).any())):
        final_games.loc[group.index, 'winning_take_foul'] = True

In [55]:
winning_take_foul = final_games[final_games['winning_take_foul'] == True]

In [56]:
check_groups(winning_take_foul)

Number of groups: 24


In [57]:
no_win_takefoul = final_games[final_games['winning_take_foul'] == False]

In [58]:
check_groups(no_win_takefoul)

Number of groups: 308


The below code goes back to the dataset analyzing take fouls when down. It creates a new column called SecsLeft that assigns how much time is remaining when the take foul actually occurs, not just when the first row is.

In [ ]:
grouped = take_foul_games.groupby('URL')

for name, group in grouped:
    if group['HomeScore'].iloc[0] > group['AwayScore'].iloc[0]:
        secs_left = group[group['HomePlay'].str.contains('take foul') & (group['SecLeft'] > 24)].SecLeft.values
        if secs_left.shape[0] > 0:
            secs_left = secs_left[0]
            take_foul_games.loc[group.index, 'SecsLeft'] = secs_left
    else:
        secs_left = group[group['AwayPlay'].str.contains('take foul') & (group['SecLeft'] > 24)].SecLeft.values
        if secs_left.shape[0] > 0:
            secs_left = secs_left[0]
            take_foul_games.loc[group.index, 'SecsLeft'] = secs_left

In [35]:
take_foul_games.to_csv('C:/Users/kevin/Downloads/take foul games.csv', index=False)
no_take_foul.to_csv('C:/Users/kevin/Downloads/no take foul games.csv', index=False)
winning_take_foul.to_csv('C:/Users/kevin/Downloads/winning take foul games.csv', index=False)
no_win_takefoul.to_csv('C:/Users/kevin/Downloads/no winning take foul games.csv', index=False)

In [36]:
take_foul_games = pd.read_csv('C:/Users/kevin/Downloads/take foul games.csv')
no_take_foul = pd.read_csv('C:/Users/kevin/Downloads/no take foul games.csv')
winning_take_foul = pd.read_csv('C:/Users/kevin/Downloads/winning take foul games.csv')
no_win_takefoul = pd.read_csv('C:/Users/kevin/Downloads/no winning take foul games.csv')

Once I have all the data I need, I simplify each game to just its first row to make it easy to analyze in part 2. Remember that the SecsLeft column shows how many seconds are left when the take foul occurs later even though the first row play is probably not the foul itself.

In [40]:
new_take_foul_games = take_foul_games.groupby('URL').first().reset_index()
new_no_take_foul = no_take_foul.groupby('URL').first().reset_index()
new_winning_take_foul = winning_take_foul.groupby('URL').first().reset_index()
new_no_win_takefoul = no_win_takefoul.groupby('URL').first().reset_index()

In [42]:
new_take_foul_games.columns

Index(['URL', 'GameType', 'Location', 'Date', 'Time', 'WinningTeam', 'Quarter',
       'SecLeft', 'AwayTeam', 'AwayPlay', 'AwayScore', 'HomeTeam', 'HomePlay',
       'HomeScore', 'Shooter', 'ShotType', 'ShotOutcome', 'ShotDist',
       'Assister', 'Blocker', 'FoulType', 'Fouler', 'Fouled', 'Rebounder',
       'ReboundType', 'ViolationPlayer', 'ViolationType', 'TimeoutTeam',
       'FreeThrowShooter', 'FreeThrowOutcome', 'FreeThrowNum', 'EnterGame',
       'LeaveGame', 'TurnoverPlayer', 'TurnoverType', 'TurnoverCause',
       'TurnoverCauser', 'JumpballAwayPlayer', 'JumpballHomePlayer',
       'JumpballPoss', 'score_diff', 'take_foul', 'SecsLeft'],
      dtype='object')

I then made a new column called 'win' in all four dataframes. For the two dataframes of analyzing the losing team, if the losing team ultimately wins the game, the 'win' column is a 1. For the two dataframes of analyzing the winning team, if the winning team ultimately wins the game, the 'win' column is a 1. 

In [61]:
new_take_foul_games['win'] = 0
new_no_take_foul['win'] = 0
new_winning_take_foul['win'] = 0
new_no_win_takefoul['win'] = 0

def check_win_losingteam(df):
    if (df['HomeScore'] < df['AwayScore']) & (df['HomeTeam'] == df['WinningTeam']):
        return 1
    elif (df['AwayScore'] < df['HomeScore']) & (df['AwayTeam'] == df['WinningTeam']):
        return 1
    else:
        return 0
    
new_take_foul_games['win'] = new_take_foul_games.apply(lambda x: check_win_losingteam(x), axis=1)
new_no_take_foul['win'] = new_no_take_foul.apply(lambda x: check_win_losingteam(x), axis=1)

def check_win_winningteam(df):
    if (df['HomeScore'] > df['AwayScore']) & (df['HomeTeam'] == df['WinningTeam']):
        return 1
    elif (df['AwayScore'] > df['HomeScore']) & (df['AwayTeam'] == df['WinningTeam']):
        return 1
    else:
        return 0

new_winning_take_foul['win'] = new_winning_take_foul.apply(lambda x: check_win_winningteam(x), axis=1)
new_no_win_takefoul['win'] = new_no_win_takefoul.apply(lambda x: check_win_winningteam(x), axis=1)

I now export these finalized dataframes to look at in part 2 - EDA

In [66]:
new_take_foul_games.to_csv('C:/Users/kevin/Downloads/simplified take foul games.csv', index=False)
new_no_take_foul.to_csv('C:/Users/kevin/Downloads/simplified no take foul games.csv', index=False)
new_winning_take_foul.to_csv('C:/Users/kevin/Downloads/simplified winning take foul games.csv', index=False)
new_no_win_takefoul.to_csv('C:/Users/kevin/Downloads/simplified no winning take foul games.csv', index=False)